client/iotcore/client.go

```diff
@@ -229,32 +229,36 @@ func (c *client) UpsertDevice(ctx context.Context,
 		return nil, err
 	}
 
-	if err == nil {
-		_, err = svc.AttachPolicy(ctx,
-			&iot.AttachPolicyInput{
-				PolicyName: aws.String(policy),
-				Target:     respCert.CertificateArn,
-			})
+	_, err = svc.AttachPolicy(ctx,
+		&iot.AttachPolicyInput{
+			PolicyName: aws.String(policy),
+			Target:     respCert.CertificateArn,
+		})
+	if err != nil {
+		return nil, fmt.Errorf("failed to attach device certificate policy: %w", err)
 	}
 
-	if err == nil {
-		_, err = svc.AttachThingPrincipal(ctx,
-			&iot.AttachThingPrincipalInput{
-				Principal: respCert.CertificateArn,
-				ThingName: aws.String(deviceID),
-			})
+	_, err = svc.AttachThingPrincipal(ctx,
+		&iot.AttachThingPrincipalInput{
+			Principal: respCert.CertificateArn,
+			ThingName: aws.String(deviceID),
+		})
+	if err != nil {
+		return nil, fmt.Errorf("failed to attach think principal: %w", err)
 	}
 
 	var deviceResp *Device
-	if err == nil {
-		deviceResp = &Device{
-			ID:          *resp.ThingId,
-			Name:        *resp.ThingName,
-			Status:      device.Status,
-			PrivateKey:  string(crypto.PrivateKeyToPem(privKey)),
-			Certificate: *respCert.CertificatePem,
-			Endpoint:    endpointOutput.EndpointAddress,
-		}
+	pkeyPEM, err := crypto.PrivateKeyToPem(privKey)
+	if err != nil {
+		return nil, fmt.Errorf("failed to serialize private key: %w", err)
+	}
+	deviceResp = &Device{
+		ID:          *resp.ThingId,
+		Name:        *resp.ThingName,
+		Status:      device.Status,
+		PrivateKey:  string(pkeyPEM),
```

Comment:
this is for my learning, why do we need to send the private key?

Reply:
This is the private key for the device to authenticate with IoT Core. It is sent as part of the configure deployment.